crunchers/cruncher_top_cards.h

```diff
@@ -0,0 +1,252 @@
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ *******************************************************************************/
+
+#ifndef CRUNCHER_TOP_CARDS_H
+#define CRUNCHER_TOP_CARDS_H
+
+#include <deque>
+#include <vector>
+#include <unordered_map>
+
+#include "../../Current/Blocks/HTTP/api.h"
+#include "../../Current/Blocks/MMQ/mmq.h"
+#include "../../Current/Bricks/util/singleton.h"
+
+#include "cruncher.h"
+#include "schema.h"
+
+namespace CTFO {
+
+CURRENT_STRUCT(CardCounters) {
+  CURRENT_FIELD(ctfo, uint64_t, 0);
+  CURRENT_FIELD(skip, uint64_t, 0);
+  CURRENT_FIELD(tfu, uint64_t, 0);
+  CURRENT_FIELD(fav, uint64_t, 0);
+  CURRENT_FIELD(unfav, uint64_t, 0);
+  CURRENT_FIELD(seen, uint64_t, 0);
+
+  bool Empty() const { return ctfo == 0 && skip == 0 && tfu == 0 && fav == 0 && unfav == 0 && seen == 0; }
+};
+
+CURRENT_STRUCT(TopCardResponseItem, CardCounters) {
+  CURRENT_FIELD(cid, uint64_t, 0);
+  CURRENT_FIELD(rate, int64_t, 0);
+};
+
+using TopCardsCruncherResponse = std::vector<TopCardResponseItem>;
+
+struct TopCardsCruncherArgs final {
+  using rate_callback_t = std::function<int64_t(const CardCounters&)>;
+
+  std::chrono::microseconds interval;
+  uint64_t top_size;
+  rate_callback_t rate_calculator;
+
+  TopCardsCruncherArgs(std::chrono::microseconds interval, uint64_t size, rate_callback_t rate_calculator)
+      : interval(interval), top_size(size), rate_calculator(rate_calculator) {}
+};
+
+template <typename NAMESPACE>
+struct TopCardsCruncherImpl {
+  using Transaction_Z = typename NAMESPACE::Transaction_T9220981828355492272;
+  using EventLogEntry = typename NAMESPACE::EventLogEntry;
+  using CardDeleted = typename NAMESPACE::Persisted_CardDeleted;
+  using iOSGenericEvent = typename NAMESPACE::iOSGenericEvent;
+  using CID = typename NAMESPACE::CID;
+  using event_t = typename NAMESPACE::CTFOLogEntry;
+  using card_t = TopCardResponseItem;
+  using value_t = TopCardsCruncherResponse;
+
+  TopCardsCruncherImpl(const TopCardsCruncherArgs& args) : args_(args) {}
+  virtual ~TopCardsCruncherImpl() = default;
+
+  void OnEvent(const event_t& e, idxts_t idxts) {
+    current_us_ = idxts.us;
+    if (Exists<Transaction_Z>(e)) {
+      const auto& transaction = Value<Transaction_Z>(e);
+      for (const auto& mutation : transaction.mutations) {
+        if (Exists<CardDeleted>(mutation)) {
+          OnCardDeleted(Value<CardDeleted>(mutation));
+        }
+      }
+    } else if (Exists<EventLogEntry>(e)) {
+      OnEventLogEntry(Value<EventLogEntry>(e));
+    }
+    while (!events_list_.empty() && events_list_.back().us + args_.interval <= current_us_) {
+      TimeWindowLeft(events_list_.back());
+      events_list_.pop_back();
```

Comment:
Maybe move it to the `TimeWindowLeft` as you have `push_front` in the `TimeWindowEntered`?